Src/DependencyCollector/Microsoft.AI.DependencyCollector/DependencyCollectorDiagnosticListener.cs

```diff
@@ -0,0 +1,252 @@
+        /// </summary>
+        public void OnCompleted()
+        {
+        }
+
+        /// <summary>
+        /// Notifies the observer that the provider has experienced an error condition.
+        /// <seealso cref="IObserver{T}.OnError(Exception)"/>
+        /// </summary>
+        /// <param name="error">An object that provides additional information about the error.</param>
+        public void OnError(Exception error)
+        {
+        }
+
+        /// <summary>
+        /// Get the DependencyTelemetry objects that are still waiting for a response from the dependency. This will most likely only be used for testing purposes.
+        /// </summary>
+        internal IEnumerable<DependencyTelemetry> PendingDependencyTelemetry
+        {
+            get { return pendingTelemetry.Values; }
+        }
+
+        /// <summary>
+        /// Diagnostic event handler method for 'System.Net.Http.Request' event.
+        /// </summary>
+        [DiagnosticName("System.Net.Http.Request")]
+        public void OnRequest(HttpRequestMessage request, Guid loggingRequestId)
+        {
+            try
+            {
+                if (request != null && request.RequestUri != null && !this.applicationInsightsUrlFilter.IsApplicationInsightsUrl(request.RequestUri.ToString()))
+                {
+                    string httpMethod = request.Method.Method;
+                    Uri requestUri = request.RequestUri;
+                    string resourceName = requestUri.AbsolutePath;
+                    if (!string.IsNullOrEmpty(httpMethod))
+                    {
+                        resourceName = httpMethod + " " + resourceName;
+                    }
+
+                    DependencyTelemetry telemetry = new DependencyTelemetry();
+                    this.client.Initialize(telemetry);
+                    telemetry.Start();
+                    telemetry.Name = resourceName;
+                    telemetry.Target = requestUri.Host;
+                    telemetry.Type = RemoteDependencyConstants.HTTP;
+                    telemetry.Data = requestUri.OriginalString;
+                    this.pendingTelemetry.TryAdd(loggingRequestId, telemetry);
+
+                    HttpRequestHeaders requestHeaders = request.Headers;
+                    if (requestHeaders != null)
+                    {
+                        if (!string.IsNullOrEmpty(telemetry.Context.InstrumentationKey) && !ContainsRequestContextKeyValue(requestHeaders, RequestResponseHeaders.RequestContextSourceKey))
+                        {
+                            string sourceApplicationId;
+                            if (this.correlationIdLookupHelper.TryGetXComponentCorrelationId(telemetry.Context.InstrumentationKey, out sourceApplicationId))
+                            {
+                                SetRequestContextKeyValue(requestHeaders, RequestResponseHeaders.RequestContextSourceKey, sourceApplicationId);
+                            }
+                        }
```

Comment:
What if the exception is thrown somewhere while trying to add cross component correlation. Why would that not allow root id / partent id headers to be set. There is room for improving some localized exception handling as well.